--- 
title: "EHH First Pass" 
author: "Nicholas F. Brazeau" 
date: "`r format(Sys.time(), '%B %d, %Y')`" 
output: 
  html_document: 
    highlight: tango 
    theme: lumen 
    toc: yes 
    toc_float: yes 
    toc_depth: 3 
    code_folding: hide
editor_options:  
  chunk_output_type: console
--- 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, fig.width = 11, fig.height = 8)
library(tidyverse)
library(gridExtra)
load("~/Documents/GitHub/mip_bigbarcode_ehh/results/01-drugres_obj_rehh.rda")
load("~/Documents/GitHub/mip_bigbarcode_ehh/results/02-drugregions_populations_subcompar.rda")
mipbb_dr_panel_vcfR <- vcfR::read.vcfR(file = "~/Documents/GitHub/mip_bigbarcode_ehh/data/polarized_mipbi_drugres_bigbarcode_panels.vcf.bgz")
mc <- readRDS("~/Documents/GitHub/mip_bigbarcode_ehh/data/monoclonals.rds")

```


## Approach

### Data Wrangling Steps

1. Convert MIPAnalyzer object to a vcfR object and then VCF
	+ Bob accepted PR, this is now in `MIPAnalyzer` as MIPanalyzerbi2vcfR. Passed `vcf-validator`
	+ One thing to consider, this function expects you to assign a "cutoff" on the Non-ref WSAF to assign the GT in the format column. This parameter will be sensitive to our results as any HET calls are coded as missing when making the haplotype files (see below) 
2. Polarized VCF (i.e. assigned ancestral allele based on _P. reichenowi_) using Andrew Morgan's `vcfdo` package. 
3. Used genetic map that Andrew Morgan put together from Pfcross to impute cM position from our VCF base-pair positions
4. Subset to monoclonal samples identified by OJ Watson (ongoing) 
5. Regions set by OJ Watson
	+ After discussion, we elected to keep this simple and use k-means clustering with a number of clusters that we wanted to see (i.e. 4 clusters in the DRC to get a N/E/S/W groups). 
6. Pulled in Drug Resistance regions identifed by Oz and converted from 0-based to 1-based. Then set window as 100kb upstream and downstream of the start/end of the gene, respectfully. 
	+ These regions are now also updated in the `rplasmodium` package 

### Data Subset and Filtering

#### Considering All Populations Together

1. Subset VCF to drug loci regions 
2. Drop any loci that does not have at **least 10 variants** in the 50kb (up/downstream) gene flanking regions
	+ This may require some tweaking testing as this is an arbitary cutoff no my end 
	+ A reasonable cutoff is needed and perhaps conditional on physical space (10 var 1 cM aren't informative but 10 var w/in 0.01 cM are) 
3. Drop loci that did not have a clear marker because of a lack of diversity as detected by `scan_hh` through the `discard_integration_at_border` flag (e.g. kelch)


#### Account for Population Substructure

1. Subset VCF to drug loci and map regions (i.e. DRC North now has a VCF for all 23 drug-resistance loci Oz identified). Calling this "region-loci" moving forward. 
2. Drop any region-loci that does not have at **least 10 variants** in the 50kb (up/downstream) gene flanking regions or 5 samples. 
	+ See above for "tweaking" 

#### `rehh` Filters

1. `rehh` package has some default fiters (see [here](https://cran.r-project.org/web/packages/rehh/vignettes/rehh.pdf) wrt to `min_perc_geno.hap`, `min_perc_geno.snp`, and `min_maf` and less wrt `limehh` and `limhaplo`).
	+ The "min" cutoffs for the `data2haplohh` function will require some "tweaking".  
	+ The defaults I used for the _haplotype statistics_ of limit haplotype and SNPs to 50% (arbitrarily) and limited ehh to 0.05 (common). 
	+ **NOTE, as a result of these `rehh` fitlers, additional samples and SNPs get dropped**. 


### Notes

vcfR manipulations are being hosted in the R-package `IDEELResearch/vcfRmanip` and includes tools to subset and alter the vcfR class. However, when writing out to disk with `vcfR::write.vcf` the compression strategy used does not agree with `bcftools` (wants `bgzip`). A quick fix is `gunzip vcf ; bgzip vcf`. 

#### The Good
When we have a site that we previously identified and sampled somewhat densely around it, the EHH plots look reasonable. 
**Differences in _crt_ and _dhps_ haplotypes by regions that correspond with Bob's phylo-maps (and Molly/Bob's previously published love-children maps).**

#### The Ugly
Kelch has very little variation -- probably just drift. DHFR wasn't sampled "densely" that passed. Pfabc is potentially noise. 

#### Table
Metadata/COI provided by OJ. Samples filtered by Bob. 
**With current filters down to 112 samples, we are left with three countries: DRC, Ghana, Zambia**.

```{r}


vcfsmpls <- unlist(colnames(mipbb_dr_panel_vcfR@gt)[-1])
mcsub <- mc[mc$ID %in% vcfsmpls, ] 
mcsub %>% 
  DT::datatable(., extensions='Buttons',
               options = list(
              searching = T,
              pageLength = 15,
              dom = 'Bfrtip', 
              buttons = c('csv')))

```

------

## Populations Considered Together

Here, we consider all population together. The _iHS_ statistic was calculated for all loci (wrt site) and the loci with the higher _iHS_ value is displayed below as an _EHH_ graph based on _a priori_ drug resistant mutations. 

### Table of Mutations 50kb Up/Downstream 
```{r}

drugres_ret_full %>% 
  dplyr::select(-c("vcfRobj")) %>% 
  DT::datatable(., extensions='Buttons',
               options = list(
              searching = T,
              pageLength = 22,
              dom = 'Bfrtip', 
              buttons = c('csv')))
```

### Plots
```{r}
plotobj <- ehhplotdf$ehhplot
n <- length(plotobj)
nCol <- floor(sqrt(n))
do.call("grid.arrange", c(plotobj, ncol=nCol))



```


## Population Structure

Here, same thing but considering pop-structure. 

### Population Map
**Based on OJ's K-means clustering with DRC of 2 groups because this was closest to SRM's question of an East/West Divide. Can split out more but waiting for filtering updates**. 

```{r, fig.width=6, fig.height=6, fig.align='center'}
basemap <- readRDS("~/Documents/GitHub/mip_bigbarcode_ehh/data/map_bases/basemap.rds")
ggplot() + 
  basemap +
  geom_point(data = mcsub, aes(x=long, y=lat, color = factor(Country))) +
  ggtitle(paste("Locations of the", nrow(mcsub), "Samples Remaining")) +
  theme(legend.position = "bottom",
        plot.title = element_text(face = "bold", hjust = 0.5),
        axis.title = element_blank(),
        axis.ticks = element_blank())

```


### Table of Mutations 50kb Up/Downstream 
```{r}

drugregions_sub %>% 
  dplyr::select(c("region", "chr", "start", "end", "name", "geneid", "nvar", "nsmpls")) %>% 
  DT::datatable(., extensions='Buttons',
               options = list(
              searching = T,
              pageLength = 22,
              dom = 'Bfrtip', 
              buttons = c('csv')))
```

### Plots
```{r, results='hide'}



final <- crossehhplotdf %>% 
  group_by(name) %>% 
  nest()
final$ehhplot <- map(final$data, "ehhplot")
final$happlot <- map(final$data, "happlot")

map2(final$ehhplot, final$happlot, function(x, y){
  n <- length(x)
  nCol <- floor(sqrt(n))
  do.call("grid.arrange", c(x, ncol=nCol))
  
  m <- length(y)
  mCol <- floor(sqrt(m))
  do.call("grid.arrange", c(y, ncol=mCol))
  
})



```
